test/verify/test_scatter0.cpp

```diff
@@ -34,7 +34,11 @@ struct test_scatter0 : verify_program<test_scatter0>
         migraphx::program p;
         auto* mm = p.get_main_module();
         migraphx::shape sd{migraphx::shape::float_type, {3, 3}};
-        migraphx::shape si{migraphx::shape::int32_type, {2, 3}};
+        // Note, scatter op should tolerate indices of type int32_t.
+        // However, the existing implementation of scatter op is
+        // not jit and is based on hip_visit_all, which requires
+        // argument types to be similar with result type.
```

Comment:
We can add int32_t to the visitor.

Reply:
It turns out that adding int32_t to the visitor doesn't help with https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/ed6542ee4ee2106fe5c104cdef7057ffdb57f314/src/targets/gpu/device/include/migraphx/gpu/device/visit.hpp#L122

But I did give a shot in just commenting it out, which results in a computing error in the scatter0 unit test. I'm not sure that's worth my time to debugging again. Could you give some context on why there's the type must be same with result requirement in the first place?

Reply:
Because they are all using the same visitor aka switch statement. If we want different types then we will need to use a nested visitor.

Reply:
Do you mean to add additional nest level in the implementation of `src/targets/gpu/device/scatter.cpp`? I have to admit I'm clueless about what is going on in that implementation or why there's already two level nest of `hip_visit_all` and `gs_launch`.

Reply:
So there is no need for us to pass `inds` to the first visitor. We can `indices.get_shape()` instead of `s1`.

But the visitors actually look wrong. The updates and data should be the same type, and only the indices are different(thats the way we implement this in the ref implementation).

Reply:
`gs_launch` will launch the lambda on the device and use a global strided loop and pass the index to the lambda(ie the `auto i` parameter).

The `hip_visit_all` will do the visit like we normally do for the data types to get a `tensor_view` but instead will create a `hip_tensor_view`, which also needs to visit the tensor rank so the dimensions and strides of the shape can be allocated as a static array(but this will make us generate a different kernel for each tensor rank).

Reply:
Hmmm... I made an attempt to change the implementation like, if that's what you suggest:

```cpp
hip_visit_all(result, arg0)([&](auto output, auto data) {
   ...
   auto s1 = arg1.get_shape();
   gs_launch...
       auto out_idx = s1.mult(i);
       ...
       out_idx[axis] = index;
```
 Then the compiler complaints that `out_idx[axis] = index` doesn't work because I'm trying to call operator[] on device. 

I don't quite get why all of the sudden this becomes an issue for the compiler... Does passing it as an argument (the original way of implementation) implicitly copy the vector to the device into a device-compatible type (which has `operator[]` implemented?

Reply:
> Then the compiler complaints that out_idx[axis] = index doesn't work because I'm trying to call operator[] on device.

The shape class doesn't run on the device, and uses dynamically allocated memory on the host. Thats why you need to pass it to the visitor to get `hip_shape` class which will run on the device and all the memory is statically allocated, so when the lambda passed to `gs_launch` captures the class by value, its copied to the device.

Reply:
> Thats why you need to pass it to the visitor to get hip_shape class which will run on the device

Are we back from where we started? Or did I captured your initial suggestion wrong?

> We can indices.get_shape() instead of s1.

From this point my understanding is that we need to get rid of this s1 argument such as to avoid the `type must match` requirement from `hip_visit_all_impl()`. But then following the conversation, getting rid of s1 won't work because it will make it the wrong `shape` type instead of `hip_shape` type. And now we are back to the need to pass the shape again to the first `hip_visit_all()`. I'm still confused about the proper way to fix the original type must match requirement.

If you have a good idea of what needs to be done, would you mind to push the commit so this becomes a collaborated effort? I think I understand why the existing kernel is implemented but to make code changes here still seems to be very tricky and it constantly results in ten pages of compiling error. Or if it is convenient for you, we can have another meeting.

Reply:
> Are we back from where we started? Or did I captured your initial suggestion wrong?

The suggestion was to use `indices.get_shape()` instead of `s1`. 

> From this point my understanding is that we need to get rid of this s1 argument such as to avoid the type must match requirement from hip_visit_all_impl(). 

This will fix the first `hip_visit_all`, but there might be a problem with the second. It should match the visits that are used in the ref implementation where `result`, `arg0`(the data), and `arg2`(the update tensor) are in the same visit and the `indices` are a seperate visit:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/op/scatter.hpp#L82